### ISSUE TITLE
Heap exhaustion consuming large request body stream

### ISSUE THREAD

**Issue by mfikes** [ztellman/aleph#230]
When reading a large HTTP request body using Aleph (with `:raw-stream? false`), I encounter heap exhaustion.

To reproduce (this is on OS X):

`lein new aleph-test` and then add `[aleph "0.4.1-beta6"]` as a `:dependency`

Then lein repl and issue the following forms:

```
(require '[aleph.http :as http]
         '[byte-streams :as bs])

(defn handler [req]
  (while (not (neg? (.read (bs/to-input-stream (:body req))))))
  {:status 400})

(http/start-server handler {:port 8081 :raw-stream? false})
```

In a separate terminal:

```
curl -F file=@file.txt http://localhost:8081
```

where `file.txt` is gigantic (sufficiently large to cause the issue to occur prior to the test completing).

I would expect this to be able to run indefinitely (with backpressure slowing down `curl` as needed, and bytes being read off the `InputStream` as fast as they can be consumed and discarded).

Ultimately you will see OOM in the REPL:

```
#object[aleph.netty$start_server$reify__6472 0x21daeb18 "aleph.netty$start_server$reify__6472@21daeb18"]
user=> Exception in thread "dirigiste-executor-controller-0" java.lang.OutOfMemoryError: Java heap space
	at java.util.concurrent.atomic.AtomicLongArray.<init>(AtomicLongArray.java:81)
	at io.aleph.dirigiste.Stats$UniformDoubleReservoir.<init>(Stats.java:68)
	at io.aleph.dirigiste.Executor.updateStats(Executor.java:376)
	at io.aleph.dirigiste.Executor.startControlLoop(Executor.java:467)
	at io.aleph.dirigiste.Executor.access$1000(Executor.java:9)
	at io.aleph.dirigiste.Executor$1.run(Executor.java:188)
	at java.lang.Thread.run(Thread.java:745)
Exception in thread "RMI TCP Connection(idle)" Exception in thread "dirigiste-executor-controller-1" java.lang.OutOfMemoryError: Java heap space
	at java.util.concurrent.atomic.AtomicLongArray.<init>(AtomicLongArray.java:81)
	at io.aleph.dirigiste.Stats$UniformDoubleReservoir.<init>(Stats.java:68)
	at io.aleph.dirigiste.Executor.updateStats(Executor.java:376)
	at io.aleph.dirigiste.Executor.startControlLoop(Executor.java:467)
	at io.aleph.dirigiste.Executor.access$1000(Executor.java:9)
	at io.aleph.dirigiste.Executor$1.run(Executor.java:188)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by ztellman** [comment]
This may have the same underlying cause as #229.  Also looking into this.

**Comment by mfikes** [comment]
Thanks! FWIW, I've also ensured that it occurs if you elide the redundant `bs/to-input-stream` coercion.

I thought it might also be related to #214, but at least with this one there is a minimal repro.

**Comment by ztellman** [comment]
Same issue as the reproducing code in #229.

**Comment by ztellman** [comment]
Wait, sorry, I missed the part where this happens even without the redundant `to-input-stream`.  Re-opening while I look into it.

**Comment by ztellman** [comment]
Okay, this revealed a fairly significant bug I introduced in byte-streams 0.2.1-alpha4, where backpressure propagation for streams wrapped in InputStreams was broken.  I appreciate the report, even if there was a bit of confusion w.r.t. the sample code.  Thanks!

**Comment by mfikes** [comment]
Excellent! I'll give the latest betas a try. In our production code (which involves raw-streams being converted to an InputStream using `to-input-stream`) we were experiencing both the heap issue and also what appears to be some sort of random livelock that does not occur with Aleph 0.4.0. 

 #229 was the result of (a botched) attempt to produce a minimal repo of the livelock issue, which involved a stack that starts off like this. If I can reproduce it or if it goes away with the latest, I'll post.

```
Name: Thread-42610
State: WAITING on java.util.concurrent.CountDownLatch$Sync@1341473
Total blocked: 0  Total waited: 17

Stack trace: 
sun.misc.Unsafe.park(Native Method)
java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
manifold.deferred.Deferred.deref(deferred.clj:439)
clojure.core$deref.invokeStatic(core.clj:2228)
clojure.core$deref.invoke(core.clj:2214)
byte_streams.pushback_stream.SynchronizedPushbackByteStream.read(pushback_stream.clj:113)
byte_streams.InputStream.read(InputStream.java:50)
sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
   - locked java.io.InputStreamReader@14903a84
java.io.InputStreamReader.read(InputStreamReader.java:184)
java.io.BufferedReader.fill(BufferedReader.java:161)
java.io.BufferedReader.read1(BufferedReader.java:212)
java.io.BufferedReader.read(BufferedReader.java:286)
   - locked java.io.InputStreamReader@14903a84
```

**Comment by ztellman** [comment]
The livelock may be related to https://github.com/ztellman/aleph/issues/193, which was recently fixed.  Hopefully `0.4.1-beta7` will be the answer to all your problems.

**Comment by mfikes** [comment]
@ztellman Just wanted to follow up with the results of using `0.4.1-beta7`: It fixes the memory leak (presumably via the revision in `[byte-streams "0.2.1-alpha1"]`)
